### ISSUE TITLE
Docker provider stops working when dockerd is restarted, requires traefik restart to fix

### ISSUE THREAD

**Issue by c2h5oh** [containous/traefik#5833]
### What did you do?

Updated Docker OS package, which caused Docker daemon (dockerd) to be restarted.
Simply restarting dockerd service has the same effect.

### What did you expect to see?

Traefik Docker provider keeps working as it was prior to dockerd restart

### What did you see instead?

Traefik Docker provider stops picking up changes until traefik is restarted.
This is likely related to https://github.com/containous/traefik/issues/5589 - dockerd socket is bind-mounted into traefik container and when dockerd is restarted that socket is recreated.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.2
Codename:     montdor
Go version:   go1.13.1
Built:        2019-10-09T19:26:05Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Traefik running in a docker container, using docker provider

```yaml
api:
  dashboard: false
  insecure: false

entryPoints:
  http:
    address: ":80"

  https:
    address: ":443"

providers:
  docker:
    endpoint: "unix:///var/run/docker.sock"
    exposedByDefault: false
```

Our dockerd config contains `"live-restore": true` which results in dockerd restarts not causing container restarts.

**Comment by sh7dm** [comment]
I tried to reproduce and it looks like it's not caused by bind-mounts. Traefik is running on host, connecting via UNIX socket and it fails when I restart Docker service.

**Comment by ginkel** [comment]
Is there a chance to at least detect this error condition via a failing health ping and/or Prometheus metric? ATM Traefik starts logging the following error, but no metrics hint at something being fundamentally broken:

```
{"level":"error","msg":"Provider connection error Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?, retrying in 9.991645219s","providerName":"docker","time":"2020-10-12T13:54:14Z"}
```

**Comment by goshander** [comment]
Same issue

**Comment by sh7dm** [comment]
Could not reproduce. Traefik at `v2.4` branch, Docker 20.10.2, no live restore.
Steps:
1. `docker run -l "traefik.http.routers.nginx.rule=Path(\"/\")" --name nginx -d nginx:alpine`
2. It works
3. `systemctl restart docker`
4. Error in Traefik log, not working (container stopped)
5. `docker start nginx`
6. Traefik picked up new data and page access works

<details>

```yaml
api:
  dashboard: true
  insecure: true

log:
  level: DEBUG

entryPoints:
  http:
    address: ":3080"

  https:
    address: ":3443"

providers:
  docker:
    endpoint: "unix:///var/run/docker.sock"
```

</details>

**Comment by c2h5oh** [comment]
> no live restore

This is the key difference. With live restore off when dockerd is restarted so is every container including traefik. Enable live restore to reproduce.

**Comment by sh7dm** [comment]
Yes, but when I explored this back in 2019 it didn't work this way too, so partially fixed (and probably the issue needs renaming).

**Comment by sh7dm** [comment]
Weird, that works as well. Maybe I have some difference in setup?

1. `docker run -l "traefik.http.routers.nginx.rule=Path(\"/\")" --name nginx -d nginx:alpine`
2. Service is available
3. `systemctl restart docker` (it does live-reload, `There are old running containers, the network config will not take affect`)
4. Error, then Docker config picked up, then service becomes available

```
ERRO[2021-06-14T16:34:04+03:00] Provider connection error unexpected EOF, retrying in 574.147282ms  providerName=docker
DEBU[2021-06-14T16:34:04+03:00] Provider connection established with docker 20.10.2 (API 1.41)  providerName=docker
DEBU[2021-06-14T16:34:04+03:00] Configuration received from provider docker: {"http":{"routers":{"nginx":{"service":"nginx","rule":"Path(\"/\")"}},"services":{"nginx":{"loadBalancer":{"servers":[{"url":"http://172.17.0.2:80"}],"passHostHeader":true}}}},"tcp":{},"udp":{}}  providerName=docker
INFO[2021-06-14T16:34:04+03:00] Skipping same configuration                   providerName=docker
```

**Comment by lifeofguenter** [comment]
Is this maybe fixable by connecting to the dockerd via IP instead of a socket file?

**Comment by c2h5oh** [comment]
> Is this maybe fixable by connecting to the dockerd via IP instead of a socket file?

Since docker daemon has not built-in authentication by doing that you are essentially giving root access to any process that can connect to localhost IP on that host.

**Comment by lifeofguenter** [comment]
> Since docker daemon has not built-in authentication by doing that you are essentially giving root access to any process that can connect to localhost IP on that host.

thats not true, you can authenticate via server-client TLS.

**Comment by michaelkebe** [comment]
We are experiencing the same issue with a setup like @jhowe-uw (VMs, docker with live-restore enabled).

We are updating docker via ansible and restarting the traefik container along with it. This is just a workaround and not very satisfactory.

We tried to use the ping healthcheck endpoint, but when it happens, the healthcheck is successful. So we are not able to detect the problem with the healthcheck.

I also think the problem is how docker handles the volume mounts.

**Comment by michaelkebe** [comment]
I investigated a little bit further. It looks like docker handles file mounts and directory mounts differently.

I think, that I found a solution:
Don't mount the `docker.sock` directly with:

    /var/run/docker.sock:/var/run/docker.sock:ro

Mount the `/var/run` directory containing the `docker.sock`

    /var/run:/var/run:ro

I cannot say much about the consequences mouting the whole `/var/run` directory.



My testsetup looks like this https://gist.github.com/michaelkebe/a1fd64c5d31aaca5b092aa2b7409bf6d
The `watcher.sh` simulates traefik trying to connect to the `docker.sock`.

1. Clone the testsetup:
`$ git clone https://gist.github.com/a1fd64c5d31aaca5b092aa2b7409bf6d.git`
2. Start the setup:
`$ cd a1fd64c5d31aaca5b092aa2b7409bf6d/`
`$ docker compose up -d`
3. Attach to the container to watch the output
`$ docker attach a1fd64c5d31aaca5b092aa2b7409bf6d-dockersockwatcher-1`
4. In another terminal run upgrades, downgrades, restarts of the docker daemon as you wish and while looking at the output of the `watcher.sh`.
`$ sudo apt install docker-ce=5:23.0.6-1~ubuntu.20.04~focal`
`$ sudo apt install docker-ce=5:24.0.6-1~ubuntu.20.04~focal`

If you want to try the different mount options, edit the `docker-compose.yml` and bring the testsetup up again.

With the not working option (mount `docker.sock` file directly) the `watcher.sh` outputs

    nc: unix connect failed: Connection refused
    nc: /var/run/docker.sock: Connection refused

**Comment by michaelkebe** [comment]
Here is a discussion exactly about this problem.

https://github.com/moby/moby/issues/22789

**Comment by Trickfilm400** [comment]
Hi together,
it's an older topic, but I have the same issue, but I came across an interesting information:

You can connect to the docker deamon via SSH:
https://doc.traefik.io/traefik/providers/docker/#endpoint
```yml
providers:
  docker:
    endpoint: "ssh://traefik@192.168.2.5:2022"
```

You need to handle the public key somehow, but maybe this is a solution? (Need to try this out soon)

**Comment by PrplHaz4** [comment]
> fixed for me by adding next items for traefik service in the `docker-compose.yaml`:
> 
> ```
>     deploy:
>       replicas: 1
>       restart_policy:
>         condition: any
> ```

Thanks for this!  I have consistently been seeing this problem when updating docker, and run most of my containers with the `unless stopped` flag.  Docker docs state for restart policy the difference from `always` - and it seems surprising that traefik would be considered `stopped` when the daemon restarts.

All my other containers come back up without issues using `unless-stopped`.  Is traefik not shutting down or restarting properly?

```
unless-stopped | Similar to always, except that when the container is stopped (manually or otherwise), it isn't restarted even after Docker daemon restarts.
```

**Comment by rgafarov** [comment]
> All my other containers come back up without issues using `unless-stopped`. Is traefik not shutting down or restarting properly?

I have no an exact idea why problem is happen, but something related to mounting docker socket as unix socket. May be after daemon restarting new inode is creating, old one is not accessible, but this is not confirmed by mounting folder instead of file. Anyhow, for some reasons traeifk container staying in Exit status after docker daemon restart. To solve this with minimal changes, without any proxy containers etc. deploy settings was added, as mentioned above.

**Comment by AlphaCraft9658** [comment]
This is something that recently started happening. It never happened before, and now I have to manually restart traefik every time to make it work again. I haven't changed anything in the configuration either.

**Comment by arp-mbender** [comment]
Having the same issue since... quite a long while ago. /var/run is... scary.

> fixed for me by adding next items for traefik service in the `docker-compose.yaml`:
> 
> ```
>     deploy:
>       replicas: 1
>       restart_policy:
>         condition: any
> ```

I think this only works if one's running Traefik as part of a swarm setup, no?
